src/lib/log.c

```diff
@@ -19,19 +19,94 @@
 
 #define _GNU_SOURCE
 
+#include <ctype.h>
 #include <stdarg.h>
+#include <stdlib.h>
+#include <string.h>
 #include <time.h>
 
 #include "log.h"
 #define DEBUG_HEADER_SIZE 50
 
 static int cur_log_level = LOG_INFO;
+static int json_format = 0;
 
 void set_log_level(int log_level)
 {
 	cur_log_level = log_level;
 }
 
+void set_json_format(void)
+{
+	json_format = 1;
+}
+
+void json_start(const char *op)
+{
+	if (json_format) {
+		fprintf(stderr, "[\n");
+		fflush(stderr);
+		fprintf(stderr, "{ \"%s\" : \"start\" },\n", op);
+		fflush(stderr);
+	}
+}
+
+void json_end(const char *op)
+{
+	if (json_format) {
+		fprintf(stderr, "{ \"%s\" : \"end\" }\n", op);
+		fflush(stderr);
+		fprintf(stderr, "]\n");
+		fflush(stderr);
+	}
+}
+
+void format_to_json(const char *type, const char *msg, va_list args_list)
+{
+	const int MAX_TYPE_LENGTH = 8;
+	char *full_msg;
+	char ltype[MAX_TYPE_LENGTH];
+
+	/* make sure the type is all lower case */
+	if (type) {
+		for (int i = 0; type[i]; i++) {
+			ltype[i] = tolower(type[i]);
+		}
+		ltype[strlen(type)] = '\0';
+	} else {
+		/* if no type was provided we default to info */
+		strcpy(ltype, "info");
+	}
+
+	/* build the full message based on all the arguments */
+	if (vasprintf(&full_msg, msg, args_list) < 0) {
+		abort();
+	}
+
+	if (strcmp(full_msg, "") != 0) {
+		int j = 0;
+		for (unsigned int i = 0; i <= strlen(full_msg); i++) {
+			/* if the message has double quotes replace them with single quotes
+			* since double quotes are not supported in JSON */
+			if (full_msg[i] == '"') {
+				full_msg[i] = '\'';
+			}
+			/* remove all '\n' from the message */
+			if (full_msg[i] != '\n') {
```

Comment:
I would replace \n with a space. So we won't connect words. "this\nword" could be "this word" instead of "thisword"

Reply:
ok, I can make that change